diff --git a/asmcomp/comballoc.ml b/asmcomp/comballoc.ml
index f03bb3e66f..510029d63c 100644
--- a/asmcomp/comballoc.ml
+++ b/asmcomp/comballoc.ml
@@ -63,7 +63,9 @@ let rec combine i allocstate =
           i.arg i.res i.dbg next, allocstate)
       end
   | Iop(Icall_ind | Icall_imm _ | Iextcall _ |
-        Itailcall_ind | Itailcall_imm _ | Ipoll _) ->
+        Itailcall_ind | Itailcall_imm _ | Ipoll _
+        | Ispecific _ (* XXX *) | Iintop Icheckbound
+        | Iintop_imm (Icheckbound, _)) ->
       let newnext = combine_restart i.next in
       (instr_cons_debug i.desc i.arg i.res i.dbg newnext,
        allocstate)
diff --git a/asmcomp/selectgen.ml b/asmcomp/selectgen.ml
index 6a56cc2a27..0c7f6e7cb1 100644
--- a/asmcomp/selectgen.ml
+++ b/asmcomp/selectgen.ml
@@ -323,12 +323,13 @@ method is_simple_expr = function
   | Cop(op, args, _) ->
       begin match op with
         (* The following may have side effects *)
-      | Capply _ | Cextcall _ | Calloc | Cstore _ | Craise _ | Copaque -> false
+      | Capply _ | Cextcall _ | Calloc | Cstore _ | Craise _ | Copaque
+      | Ccheckbound -> false
         (* The remaining operations are simple if their args are *)
       | Cload _ | Caddi | Csubi | Cmuli | Cmulhi | Cdivi | Cmodi | Cand | Cor
       | Cxor | Clsl | Clsr | Casr | Ccmpi _ | Caddv | Cadda | Ccmpa _ | Cnegf
       | Cabsf | Caddf | Csubf | Cmulf | Cdivf | Cfloatofint | Cintoffloat
-      | Ccmpf _ | Ccheckbound -> List.for_all self#is_simple_expr args
+      | Ccmpf _ -> List.for_all self#is_simple_expr args
       end
   | Cassign _ | Cifthenelse _ | Cswitch _ | Ccatch _ | Cexit _
   | Ctrywith _ -> false
diff --git a/driver/compenv.ml b/driver/compenv.ml
index 8b1c4e5559..3e7de2533d 100644
--- a/driver/compenv.ml
+++ b/driver/compenv.ml
@@ -489,6 +489,7 @@ let read_one_param ppf position name v =
 let read_OCAMLPARAM ppf position =
   try
     let s = Sys.getenv "OCAMLPARAM" in
+    Warnings.parsed_ocamlparam := s;
     if s <> "" then
       let (before, after) =
         try
diff --git a/stdlib/bytes.ml b/stdlib/bytes.ml
index ae946dfa5e..319155a36c 100644
--- a/stdlib/bytes.ml
+++ b/stdlib/bytes.ml
@@ -62,8 +62,11 @@ let to_string b = unsafe_to_string (copy b)
 let of_string s = copy (unsafe_of_string s)
 
 let sub s ofs len =
-  if ofs < 0 || len < 0 || ofs > length s - len
-  then invalid_arg "String.sub / Bytes.sub"
+  let length_s = length s in
+  if ofs < 0 || len < 0 || ofs > length_s - len
+  then invalid_arg ("String.sub / Bytes.sub: ofs=" ^ string_of_int ofs
+                    ^ ", len=" ^ string_of_int len
+                    ^ ", length_s=" ^ string_of_int length_s)
   else begin
     let r = create len in
     unsafe_blit s ofs r 0 len;
diff --git a/utils/warnings.ml b/utils/warnings.ml
index 895ef2be07..3785f61590 100644
--- a/utils/warnings.ml
+++ b/utils/warnings.ml
@@ -459,7 +459,14 @@ let name_to_number =
 
 (* Must be the max number returned by the [number] function. *)
 
-let letter = function
+let parsed_ocamlparam = ref "<not-set>"
+
+let last_l : int ref = ref 0
+
+external abort : unit -> unit = "caml_gc_stat" "abort" [@@noalloc]
+
+(* CR-soon xclerc for xclerc: remove the `for_debug` parameter... *)
+let letter for_debug = function
   | 'a' ->
      let rec loop i = if i = 0 then [] else i :: loop (i - 1) in
      loop last_warning_number
@@ -488,7 +495,11 @@ let letter = function
   | 'x' -> [14; 15; 16; 17; 18; 19; 20; 21; 22; 23; 24; 30]
   | 'y' -> [26]
   | 'z' -> [27]
-  | _ -> assert false
+  | chr ->
+     let ocamlparam_from_env = match Sys.getenv_opt "OCAMLPARAM" with None -> "-" | Some  value -> value in
+     Printf.eprintf "Warnings.letter %C (last_l=%d, for_debug=%S, ocamlparam_from_env=%S ocamlparam_from_compenv=%S)\n%!" chr !last_l for_debug ocamlparam_from_env !parsed_ocamlparam;
+     abort ();
+     assert false
 ;;
 
 type state =
@@ -738,14 +749,16 @@ let parse_opt error active errflag s =
           error.(i) <- true
         end
   in
-  let eval = function
+  let eval l =
+    last_l := ((Obj.magic l) : int);
+    match l with
     | Letter(c, m) ->
         let lc = Char.lowercase_ascii c in
         let modifier = match m with
           | None -> if c = lc then Clear else Set
           | Some m -> m
         in
-        List.iter (action modifier) (letter lc)
+        List.iter (action modifier) (letter s lc)
     | Num(n1,n2,modifier) ->
         for n = n1 to Int.min n2 last_warning_number do action modifier n done
   in
@@ -1131,7 +1144,7 @@ let help_warnings () =
   print_endline "  A all warnings";
   for i = Char.code 'b' to Char.code 'z' do
     let c = Char.chr i in
-    match letter c with
+    match letter "<help-warnings>" c with
     | [] -> ()
     | [n] ->
         Printf.printf "  %c Alias for warning %i.\n" (Char.uppercase_ascii c) n
diff --git a/utils/warnings.mli b/utils/warnings.mli
index 3d9ea91f38..70d17a9181 100644
--- a/utils/warnings.mli
+++ b/utils/warnings.mli
@@ -161,3 +161,5 @@ type description =
     description : string; }
 
 val descriptions : description list
+
+val parsed_ocamlparam : string ref
